### PR TITLE
Add PauliList class

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Add `PauliList` to replace `PauliTable` and `StabilizerTable`

### Details and comments

This class functions as an optimized list of the new `Pauli` class (including phase).

- [x] tests
- [x] Improve documentation

### CODE REVIEW

File: qiskit/quantum_info/operators/symplectic/pauli_list.py
Comment (ShellyGarion) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:417]: is this TODO still needed?

File: qiskit/quantum_info/operators/symplectic/pauli_list.py
Comment (ShellyGarion) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:886]: can you maybe do it directly without decomposing the cliffolrd? like its done in #6297

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:178]: In the init tests - maybe worth checking more phases with `i` and `-i`

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:284]: can test also pseudo-random pauli's

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:368]: Z appears twice

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:409]: Z appears twice

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:423]: can try more phases as well as pseudo-random paulis

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:559]: in operators tests can test more phases and also pseudo-random paulis

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:784]: in sorting tests can test more phases

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:1040]: in delete and insert tests - maybe add phases?

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ShellyGarion) [inline test/python/quantum_info/operators/symplectic/test_pauli_list.py:1178]: in commute / anti-commute tests - maybe add phases?

File: qiskit/quantum_info/operators/symplectic/pauli_list.py
Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:417]: Thank you for your pointing out. I forget fixing the phase.

File: qiskit/quantum_info/operators/symplectic/pauli_list.py
Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:886]: Thanks. I've moved the _evolve_clifford to the base class.

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:178]: Thanks. I added the tests.

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:284]: Thanks. Why do we need the test for the random Paulis here?

File: test/python/quantum_info/operators/symplectic/test_pauli_list.py
Comment (ikkoham) [reply test/python/quantum_info/operators/symplectic/test_pauli_list.py:368]: This is as intended. We also see that the same Pauli is not omitted.

**Comment by ikkoham** [comment]
@ShellyGarion Thank you very much for your review. The missing tests about the phase were fatal, and I forgot the TODO. I am thankful for your review helped me realize them.

I have fixed them in the PR: https://github.com/chriseclectic/qiskit-terra/pull/24

File: qiskit/quantum_info/operators/symplectic/pauli_list.py
Comment (t-imamichi) [inline qiskit/quantum_info/operators/symplectic/pauli_list.py:192]: According to python docs, a form "<...>" is suggested as follows. How about changing prefix to "<PauliList("?
> Called by the repr() built-in function to compute the “official” string representation of an object. If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned.
https://docs.python.org/3/reference/datamodel.html#object.__repr__

Comment (ikkoham) [reply qiskit/quantum_info/operators/symplectic/pauli_list.py:192]: Thank you for your review. `PauliList.__repr__` is evaluatable unless it is truncated. So, this is valid Python expression.

**Review by ShellyGarion — APPROVED** [review]

**Review by t-imamichi — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]
LGTM

**Review by mtreinish — CHANGES_REQUESTED** [review]
Actually I missed it initially, this needs a release note documenting the new class. After that's added this is ready to go

**Comment by ikkoham** [comment]
@mtreinish Thank you. I made a PR to add the release note. https://github.com/chriseclectic/qiskit-terra/pull/25

@chriseclectic Could you merge this PR if it's OK.

**Review by mtreinish — APPROVED** [review]
LGTM, I just merged @ikkoham's commit with the release notes from https://github.com/chriseclectic/qiskit-terra/pull/25

**Comment by mtreinish** [comment]
It looks like there is a conflict here now: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=29265&view=logs&j=377f2c8d-8b6f-583f-e152-2ff025e2543a&t=28b3ff0d-9e76-5e94-89e9-eed7321a383e&l=12999 probably after #6442 merged

**Comment by ikkoham** [comment]
OK. I'll fix this soon.

**Comment by ikkoham** [comment]
Thank you. I think this PR fixes the error on CI. (I've tested on my local and it passed.)
https://github.com/chriseclectic/qiskit-terra/pull/26

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
